components/support/sync15-traits/src/bridged_engine.rs

```diff
@@ -84,28 +90,110 @@ pub trait BridgedEngine {
+    pub envelopes: Vec<OutgoingEnvelope>,
     /// The number of incoming records whose contents were merged because they
     /// changed on both sides. None indicates we aren't reporting this
     /// information.
     pub num_reconciled: Option<usize>,
 }
 
 impl ApplyResults {
-    pub fn new(records: Vec<String>, num_reconciled: impl Into<Option<usize>>) -> Self {
+    pub fn new(envelopes: Vec<OutgoingEnvelope>, num_reconciled: impl Into<Option<usize>>) -> Self {
         Self {
-            records,
+            envelopes,
             num_reconciled: num_reconciled.into(),
         }
     }
 }
 
 // Shorthand for engines that don't care.
-impl From<Vec<String>> for ApplyResults {
-    fn from(records: Vec<String>) -> Self {
+impl From<Vec<OutgoingEnvelope>> for ApplyResults {
+    fn from(envelopes: Vec<OutgoingEnvelope>) -> Self {
         Self {
-            records,
+            envelopes,
             num_reconciled: None,
         }
     }
 }
+
+/// An envelope for an incoming item, passed to `BridgedEngine::store_incoming`.
+/// Envelopes are a halfway point between BSOs, the format used for all items on
+/// the Sync server, and records, which are specific to each engine.
+///
+/// A BSO is a JSON object with metadata fields (`id`, `modifed`, `sortindex`),
+/// and a BSO payload that is itself a JSON string. For encrypted records, the
+/// BSO payload has a ciphertext, which must be decrypted to yield a cleartext.
+/// The cleartext is a JSON string (that's three levels of JSON wrapping, if
+/// you're keeping score: the BSO itself, BSO payload, and cleartext) with the
+/// actual record payload.
+///
+/// An envelope combines the metadata fields from the BSO, and the cleartext
+/// from the encrypted BSO payload.
+#[derive(Clone, Debug, Deserialize)]
+pub struct IncomingEnvelope {
+    pub id: Guid,
+    pub modified: ServerTimestamp,
+    #[serde(default)]
+    pub sortindex: Option<i32>,
+    // Don't provide access to the cleartext directly. We want all callers to
+    // use `IncomingEnvelope::payload`, so that we can validate the cleartext.
+    cleartext: String,
+}
+
+impl IncomingEnvelope {
+    /// Parses and returns the record payload from this envelope. Returns an
+    /// error if the envelope's cleartext isn't valid JSON, or the payload is
+    /// invalid.
+    pub fn payload(&self) -> Result<Payload, Box<dyn Error>> {
+        let payload: Payload = serde_json::from_str(&self.cleartext)?;
+        if payload.id != self.id {
+            return Err(MismatchedIdError {
+                envelope: self.id.clone(),
+                payload: payload.id,
+            }
+            .into());
+        }
+        Ok(payload)
+    }
+}
+
+/// An envelope for an outgoing item, returned from `BridgedEngine::apply`. This
+/// is similar to `IncomingEnvelope`, but omits fields that are only set by the
+/// server, like `modified`.
+#[derive(Clone, Debug, Serialize)]
+pub struct OutgoingEnvelope {
+    id: Guid,
```

Comment:
shouldn't `sortindex` be here too? (Not that we use it in rust yet, so I'd be fine with a "todo" style comment, but IIUC, conceptually it should exist here, right?)